File: xrpl4j-android/src/main/java/app/girin/xrpl4j/client/JsonParser.java
package app.girin.xrpl4j.client;

import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.xrpl.xrpl4j.model.client.XrplResult;
import org.xrpl.xrpl4j.model.jackson.ObjectMapperFactory;

public class JsonParser {
  private static final ObjectMapper objectMapper = ObjectMapperFactory.create();

  public static <T extends XrplResult> T parseResponse(JsonNode json, Class<T> resultType) throws Exception {
    try {
      JavaType javaType = objectMapper.constructType(resultType);
      return parseResponse(json, javaType);
    } catch (Exception e) {
      throw new Exception(e);
    }
  }

  public static <T extends XrplResult> T parseResponse(JsonNode json, JavaType javaType) throws Exception {
    JsonNode resultNode = json.get("result");
    try {
      return objectMapper.readValue(resultNode.toString(), javaType);
    } catch (Exception e) {
      throw new Exception(e);
    }
  }
}

File: xrpl4j-android/src/main/java/app/girin/xrpl4j/client/JsonParser.kt
package app.girin.xrpl4j.client

import com.fasterxml.jackson.databind.JavaType
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import org.xrpl.xrpl4j.model.client.XrplResult
import org.xrpl.xrpl4j.model.jackson.ObjectMapperFactory

object JsonParser {
  private val objectMapper: ObjectMapper = ObjectMapperFactory.create()

  @Throws(Exception::class)
  fun <T : XrplResult> parseResponse(json: JsonNode, resultType: Class<T>): T {
    return try {
      val javaType = objectMapper.constructType(resultType)
      parseResponse(json, javaType)
    } catch (e: Exception) {
      throw Exception(e)
    }
  }

  @Throws(Exception::class)
  fun <T : XrplResult> parseResponse(json: JsonNode, javaType: JavaType): T {
    val resultNode = json.get("result")
    return try {
      objectMapper.readValue(resultNode.toString(), javaType)
    } catch (e: Exception) {
      throw Exception(e)
    }
  }
}